### PR TITLE
Workflow now decorates cases with extra metadata when possible (subject & establishment)

### PR DESCRIPTION
**Pull request by wheelsandcogs**
When a new case is created, workflow will add a `subject` (profile id that the case is related to / owned by) and if sensible, an `establishmentId`.

### CODE REVIEW

File: lib/api.js [hunk 1]
```diff
@@ -12,6 +13,8 @@ module.exports = settings => {
 
   const flow = Taskflow({ db: settings.taskflowDB });
 
+  flow.hook('pre-create', hooks.validate(settings));
```
Comment (wheelsandcogs) [inline lib/api.js]: Is `pre-create` the right event?

**Comment by wheelsandcogs** [comment]
Open to suggestions for a better hook name than `meta`?

File: lib/hooks/meta/index.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  const { Roles } = db(settings.db);
+
+  return model => {
+    const type = get(model, 'data.model');
+    const data = get(model, 'data');
+
+    switch (type) {
+      case 'pil':
+        data.subject = get(data, 'data.profileId'); // pil holder
+        data.establishmentId = get(data, 'data.establishmentId');
```
Comment (lennym) [inline lib/hooks/meta/index.js]: I don't think this will be sufficient. Those parameters won't be on all cases, only the ones we have now with a `create` action.

I think we need to load the data from the database based on the id provided (if an id is provided).

File: lib/hooks/meta/index.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        break;
+
+      case 'place':
+        return Promise.resolve()
+          .then(() => Roles.query().findOne({ type: 'pelh', establishmentId: get(data, 'data.establishmentId') }))
+          .then(role => {
+            data.subject = role.profileId; // establishment licence holder
+            data.establishmentId = get(data, 'data.establishmentId');
+          })
+          .then(() => model.update(data));
```
Comment (lennym) [inline lib/hooks/meta/index.js]: As above, the `establishmentId` _might_ not be on the payload if it's an amendment.

**Comment by joefitter** [comment]
LGTM, I'd like to defer to @lennym though as workflow is his baby

File: lib/hooks/meta/index.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        data.subject = get(data, 'data.id'); // profile user
+        break;
+
+      // case 'project':
+      //   data.subject = ''; // project licence holder
+      //   data.establishmentId = get(data, 'data.establishmentId');
+      //   break;
+
+      case 'trainingModule':
+        data.subject = get(data, 'data.profileId');
```
Comment (lennym) [inline lib/hooks/meta/index.js]: I don't think we'll _ever_ need metadata on a training module since they'll always insta-resolve.

*If* we decide we do want metadata then the same logic as above will apply.

**Comment by lennym** [comment]
tl;dr - it's assuming params are always present on the payload. We should read associations from the db instead.

**Comment by lennym** [comment]
Hook name is fine by me.
